src/support/small_set.h

```diff
@@ -39,16 +40,19 @@ template<typename T, size_t N, typename FlexibleSet> class SmallSetBase {
   size_t usedFixed = 0;
   std::array<T, N> fixed;
 
-  // flexible additional storage
-  FlexibleSet flexible;
+  // flexible additional storage.
+  // Because std::set allocates the root node of its red-black tree on the
+  // heap, we wrap it std::optional to defer that allocation until we start
+  // using the flexible storage.
+  std::optional<FlexibleSet> flexible;
 
   bool usingFixed() const {
-    // If the flexible storage contains something, then we are using it.
+    // If the flexible optional has a value, we are using it.
     // Otherwise we use the fixed storage. Note that if we grow and shrink then
     // we will stay in flexible mode until we reach a size of zero, at which
     // point we return to fixed mode. This is intentional, to avoid a lot of
     // movement in switching between fixed and flexible mode.
-    return flexible.empty();
+    return !flexible.has_value();
```

Comment:
I'm not an expert on `std::optional`, but wouldn't this work?
```suggestion
    return !flexible;
```

Reply:
Yes, that works, too 👍